### PR TITLE
CHT test: Present 2 chains after the checkpoint - issue 40

### PR DESCRIPTION
**Pull request by fenix2222**
CT advances after last checkpoint to height LC + MaxReorg + 10. New chain is presented with fork point at LC + 5, chain is not accepted.

See #1321 (comment)

Preferred reviewers: @Aprogiena or @noescape00

### CODE REVIEW

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 1]
```diff
@@ -1552,5 +1552,64 @@ public void ChainTipAdvancesAt150Blocks_AlternativeChainPresented_RelevantBlockD
...
+            ChainedHeaderTree cht = ctx.ChainedHeaderTree;
+            ChainedHeader chainTip = ctx.InitialChainTip;
+
+            const int checkpointHeight = 5;
+            var checkpoint = new CheckpointFixture(checkpointHeight, chainTip.Header);
+            ctx.SetupCheckpoints(checkpoint);
+
+            // Setup max reorg to 100.
+            const int maxReorg = 100;
+            ctx.ChainStateMock.Setup(x => x.MaxReorgLength).Returns(maxReorg);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: You also need to set up finalised block.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 2]
```diff
@@ -1552,5 +1552,64 @@ public void ChainTipAdvancesAt150Blocks_AlternativeChainPresented_RelevantBlockD
...
+            const int chainBSize = 20;
+            ChainedHeader forkTip = chainATip.GetAncestor(checkpointHeight + forkPointHeight);
+            ChainedHeader chainBTip = ctx.ExtendAChain(initialChainSize + chainBSize - forkPointHeight, forkTip);
+
+            // Chain B is presented by peer 2.
+            List<BlockHeader> listOfChainBHeaders = ctx.ChainedHeaderToList(chainBTip, chainBSize);
+            ConnectNewHeadersResult connectNewHeadersResult = cht.ConnectNewHeaders(2, listOfChainBHeaders);
+
+            // Nothing is marked for download.
+            connectNewHeadersResult.DownloadFrom.Should().BeNull();
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: The testable outcome here should be exception when the headers are attempted to be connected.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs (hunk 1 above)
Comment (fenix2222) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Done

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 3]
```diff
@@ -1652,5 +1652,64 @@ public void ChainWith10DataAvailableHeaders_PartialValidationSucceededOnTheFirst
...
+                cht.BlockDataDownloaded(currentChainTip, block);
+                cht.PartialValidationSucceeded(currentChainTip, out bool fullValidationRequired);
+                ctx.FinalizedBlockMock.Setup(m => m.GetFinalizedBlockHeight()).Returns(currentChainTip.Height - maxReorg);
+                cht.ConsensusTipChanged(currentChainTip);
+            }
+
+            // Create new chain B with 20 headers and a fork point at height 10.
+            const int forkPointHeight = 10;
+            const int chainBSize = 20;
+            ChainedHeader forkTip = chainATip.GetAncestor(checkpointHeight + forkPointHeight);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: This should be only fork point height, or FP height should be 5

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 4]
```diff
@@ -1652,5 +1652,64 @@ public void ChainWith10DataAvailableHeaders_PartialValidationSucceededOnTheFirst
...
+            ChainedHeader chainBTip = ctx.ExtendAChain(initialChainSize + chainBSize - forkPointHeight, forkTip);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: quite confusing this

imho

fpH is 10
forkTip should be GA(fpH) only
cBTip should be EC(cBSize - fpH)

**Comment by Aprogiena** [comment]
Looks good, I think.

**Comment by noescape00** [comment]
Conflicts

**Comment by noescape00** [comment]
Conflicts again

**Pull request merged**